File: ProjectEuler/Problem100/ArrangedProbability.hs
import Control.Monad

check n = let b = ceiling ((fromIntegral n) / (sqrt 2))
          in 2*b*(b-1) == n*(n-1)

solve n = do
  n' <- filter check . filter (\m -> m `rem` 4 == 0 || m `rem` 4 == 1) $ [n..]
  return $ ceiling $ (fromIntegral n') / (sqrt 2)
  
main = print $ head $ solve n
  where n = 10000000

File: ProjectEuler/Problem100/ArrangedProbability.cpp
#include <stdio.h>
#include <math.h>

bool check(unsigned long n) {
	unsigned long b = n / sqrt(2) + 1;
	return (2*b*(b-1) == n*(n-1));
}

int main(void) {
	unsigned long i;

	for (i = (unsigned long)pow(10,12); ;) {
		if (check(i))
			break;
		i += 1 + 2*(i%4);
	}
	printf("%lu\n",(unsigned long)(i / sqrt(2) + 1));
	
	return 0;
}
